File: src/core/net/strategies/switch-account.test.ts
import { describe, expect, it } from 'vitest'
import type { NgaCredentials } from '../auth'
import { createComboCache } from '../combo'
import { createNgaFetcher } from '../fetcher'
import type { HttpRequest, HttpResponse, HttpTransport } from '../transport'
import { createFormatRotationStrategy } from './format-rotation'
import { createSwitchAccountStrategy, nextCredentialsAfter } from './switch-account'

const utf8 = (text: string) => new TextEncoder().encode(text)

const OK = utf8('{"data":{"0":"ok"},"time":1}')
const BLOCKED = utf8('<html><body>403 Forbidden</body></html>')

const ALICE: NgaCredentials = { uid: '10000001', token: 'cid-a' }
const BOB: NgaCredentials = { uid: '10000002', token: 'cid-b' }

function fakeTransport(
  respond: (request: HttpRequest) => Partial<HttpResponse>,
): { transport: HttpTransport; requests: HttpRequest[] } {
  const requests: HttpRequest[] = []
  const transport: HttpTransport = (request) => {
    requests.push(request)
    const response = respond(request)
    return Promise.resolve({
      status: response.status ?? 200,
      contentType: response.contentType ?? 'text/javascript; charset=UTF-8',
      body: response.body ?? OK,
    })
  }
  return { transport, requests }
}

const uidOf = (request: HttpRequest) =>
  /ngaPassportUid=(\d+)/.exec(request.headers.Cookie ?? '')?.[1] ?? null

/** 只有 Bob 的 cookie 能拿到数据，Alice 一律被封。 */
const onlyBobWorks = () =>
  fakeTransport((request) =>
    uidOf(request) === BOB.uid
      ? { body: OK }
      : { status: 403, body: BLOCKED, contentType: 'text/html' },
  )

function chainFetcher(transport: HttpTransport, accounts: readonly NgaCredentials[]) {
  return createNgaFetcher({
    transport,
    comboCache: createComboCache(),
    getCredentials: () => accounts[0] ?? null,
    strategies: [
      createFormatRotationStrategy({ formats: ['json'], hosts: ['https://bbs.nga.cn'] }),
      createSwitchAccountStrategy({ listCredentials: () => accounts }),
    ],
  })
}

describe('nextCredentialsAfter', () => {
  it('取当前账号之后的下一个,循环', () => {
    expect(nextCredentialsAfter([ALICE, BOB], ALICE)).toEqual(BOB)
    expect(nextCredentialsAfter([ALICE, BOB], BOB)).toEqual(ALICE)
  })

  it('不足两个账号就没得换', () => {
    expect(nextCredentialsAfter([], null)).toBeNull()
    expect(nextCredentialsAfter([ALICE], ALICE)).toBeNull()
  })

  it('当前是游客(或刚退出登录)时从头一个开始', () => {
    expect(nextCredentialsAfter([ALICE, BOB], null)).toEqual(ALICE)
    expect(nextCredentialsAfter([ALICE, BOB], { uid: '999', token: 'x' })).toEqual(ALICE)
  })
})

describe('createSwitchAccountStrategy · 换账号重试', () => {
  it('多账号时取下一个账号的 cookie 重试,成了就用它的结果', async () => {
    const { transport, requests } = onlyBobWorks()

    const result = await chainFetcher(transport, [ALICE, BOB])({ path: 'thread.php' })

    expect(result.via).toBe('switch-account')
    expect(requests.map(uidOf)).toEqual([ALICE.uid, BOB.uid])
  })

  it('只有一个账号时这一档不启用,一次请求都不发', async () => {
    const { transport, requests } = onlyBobWorks()

    await expect(chainFetcher(transport, [ALICE])({ path: 'thread.php' })).rejects.toMatchObject({
      kind: 'parse',
    })

    // 只有 format-rotation 那一档发过请求,全是 Alice
    expect(requests.map(uidOf)).toEqual([ALICE.uid])
  })

  it('只试一次:换的那个账号也被封就交给链上后面的兜底,不再换第三个', async () => {
    const CAROL: NgaCredentials = { uid: '10000003', token: 'cid-c' }
    const { transport, requests } = fakeTransport(() => ({
      status: 403,
      body: BLOCKED,
      contentType: 'text/html',
    }))

    await expect(
      chainFetcher(transport, [ALICE, BOB, CAROL])({ path: 'thread.php' }),
    ).rejects.toMatchObject({ kind: 'parse' })

    expect(requests.map(uidOf)).toEqual([ALICE.uid, BOB.uid])
  })

  it('沿用缓存里的成功组合:这一档变的是身份,不是组合', async () => {
    const cache = createComboCache()
    cache.remember('thread.php', { format: 'jsonLite', host: 'https://ngabbs.com' })
    const { transport, requests } = onlyBobWorks()
    const fetchNga = createNgaFetcher({
      transport,
      comboCache: cache,
      getCredentials: () => ALICE,
      strategies: [createSwitchAccountStrategy({ listCredentials: () => [ALICE, BOB] })],
    })

    await fetchNga({ path: 'thread.php' })

    expect(requests).toHaveLength(1)
    expect(requests[0]!.url).toContain('https://ngabbs.com/thread.php')
    expect(requests[0]!.url).toContain('lite=js')
  })

  it('业务错误照样不重试:换个账号也还是同一个语义错误', async () => {
    const { transport, requests } = fakeTransport(() => ({
      body: utf8('{"error":{"0":"2048:找不到主题"}}'),
    }))

    await expect(
      chainFetcher(transport, [ALICE, BOB])({ path: 'read.php', query: { tid: 1 } }),
    ).rejects.toMatchObject({ kind: 'server' })

    expect(requests).toHaveLength(1)
  })
})

File: native/app/src/test/kotlin/com/chasel/ng2n/core/net/SwitchAccountTest.kt
package com.chasel.ng2n.core.net

import com.chasel.ng2n.core.net.strategies.FormatRotationStrategy
import com.chasel.ng2n.core.net.strategies.SwitchAccountStrategy
import com.chasel.ng2n.core.net.strategies.nextCredentialAfter
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * 逐条移植自 `src/core/net/strategies/switch-account.test.ts`
 * (`nextCredentialsAfter` 3 条 + `换账号重试` 5 条 = 8 条,全部移植)。
 */
class SwitchAccountTest {

  private fun uidOf(request: HttpRequest) = request.credential?.uid

  /** 只有 Bob 的凭证能拿到数据,Alice 一律被封。 */
  private fun onlyBobWorks() =
    RecordingTransport { if (uidOf(it) == BOB.uid) ok() else blocked() }

  private fun chainClient(
    transport: Transport,
    accounts: List<Credential>,
    cache: ComboCache = InMemoryComboCache(),
  ) = testClient(
    transport,
    strategies = listOf(
      FormatRotationStrategy(listOf(ResponseFormat.JSON), listOf("https://bbs.nga.cn")),
      SwitchAccountStrategy { accounts },
    ),
    credentials = FakeCredentials(signedIn = accounts.firstOrNull(), accounts = accounts),
    comboCache = cache,
  )

  // ── nextCredentialAfter ───────────────────────────────────────────────────

  @Test
  fun `取当前账号之后的下一个,循环`() {
    assertEquals(BOB, nextCredentialAfter(listOf(ALICE, BOB), ALICE))
    assertEquals(ALICE, nextCredentialAfter(listOf(ALICE, BOB), BOB))
  }

  @Test
  fun `不足两个账号就没得换`() {
    assertNull(nextCredentialAfter(emptyList(), null))
    assertNull(nextCredentialAfter(listOf(ALICE), ALICE))
  }

  @Test
  fun `当前是游客(或刚退出登录)时从头一个开始`() {
    assertEquals(ALICE, nextCredentialAfter(listOf(ALICE, BOB), null))
    assertEquals(ALICE, nextCredentialAfter(listOf(ALICE, BOB), Credential("999", "x")))
  }

  // ── 换账号重试 ─────────────────────────────────────────────────────────────

  @Test
  fun `多账号时取下一个账号的 cookie 重试,成了就用它的结果`() = runTest {
    val transport = onlyBobWorks()

    val result = chainClient(transport, listOf(ALICE, BOB)).execute(readRequest("thread.php"))

    assertEquals("switch-account", result.via)
    assertEquals(listOf(ALICE.uid, BOB.uid), transport.uids())
  }

  @Test
  fun `只有一个账号时这一档不启用,一次请求都不发`() = runTest {
    val transport = onlyBobWorks()

    val error = assertThrowsNga {
      chainClient(transport, listOf(ALICE)).execute(readRequest("thread.php"))
    }

    assertEquals(NgaErrorKind.PARSE, error.kind)
    // 只有 format-rotation 那一档发过请求,全是 Alice
    assertEquals(listOf(ALICE.uid), transport.uids())
  }

  @Test
  fun `只试一次·换的那个账号也被封就交给链上后面的兜底,不再换第三个`() = runTest {
    val transport = RecordingTransport { blocked() }

    val error = assertThrowsNga {
      chainClient(transport, listOf(ALICE, BOB, CAROL)).execute(readRequest("thread.php"))
    }

    assertEquals(NgaErrorKind.PARSE, error.kind)
    assertEquals(listOf(ALICE.uid, BOB.uid), transport.uids())
  }

  @Test
  fun `沿用缓存里的成功组合·这一档变的是身份,不是组合`() = runTest {
    val cache = InMemoryComboCache()
    cache.remember("thread.php", FetchCombo(ResponseFormat.JSON_LITE, "https://ngabbs.com"))
    val transport = onlyBobWorks()
    val client = testClient(
      transport,
      strategies = listOf(SwitchAccountStrategy { listOf(ALICE, BOB) }),
      credentials = FakeCredentials(signedIn = ALICE, accounts = listOf(ALICE, BOB)),
      comboCache = cache,
    )

    client.execute(readRequest("thread.php"))

    assertEquals(1, transport.requests.size)
    assertTrue(transport.requests[0].url.startsWith("https://ngabbs.com/thread.php"))
    assertTrue(transport.requests[0].url.contains("lite=js"))
  }

  @Test
  fun `业务错误照样不重试·换个账号也还是同一个语义错误`() = runTest {
    val transport = RecordingTransport { ok("""{"error":{"0":"2048:找不到主题"}}""") }

    val error = assertThrowsNga {
      chainClient(transport, listOf(ALICE, BOB))
        .execute(readRequest("read.php", queryOf("tid" to 1)))
    }

    assertEquals(NgaErrorKind.SERVER, error.kind)
    assertEquals(1, transport.requests.size)
  }
}
